Allow an op to be falsy

antoinelyset opened the pull request:
Hi!

This is a fix for https://github.com/share/sharedb/issues/373

I chose to use `'op' in op` as the check to know if we are doing an edit.

I've added a simple test to verify that it works (it was actually really useful to find all the places I needed to change).

curran commented on lib/agent.js:
```diff
@@ -257,7 +257,7 @@ Agent.prototype._sendOp = function(collection, id, op) {
     seq: op.seq
   };
-  if (op.op) message.op = op.op;
+  if ('op' in op) message.op = op.op;
```
Wait, so `op` here can be `null`, right?

But, `in` doesn't work on `null`...

![image](https://user-images.githubusercontent.com/68416/85033623-2a64c000-b14f-11ea-8bc7-83f5f55113d7.png)

In any case, since the solution will need to be duplicated so many places across the codebase, it might make sense to introduce a helper function, like `hasOp(op)`, so that the actual checking logic could be modified in just one place.

antoinelyset replied:
Here `op` will always be an object I think. (The old condition was based on this too). So more

<img width="555" alt="Screenshot 2020-06-18 at 16 43 51" src="https://user-images.githubusercontent.com/1465371/85034948-ec818f00-b182-11ea-87c5-3112d487798e.png">

Calyhre replied:
Yep, @antoinelyset is right here, `op` is an object. It's `op.op` in this code that might be `null`

alecgibson commented on lib/client/doc.js:
```diff
@@ -528,7 +528,7 @@ function transformX(client, server) {
 
   // Ignore no-op coming from server
-  if (!server.op) return;
+  if (!('op' in server)) return;
```
As I mentioned in the issue, can we please update `setNoOp` to return a no-op consistent with this new shape (ie `{op: null}`), and then keep this check as it was before.

antoinelyset replied:
Sorry, I'm a bit confused by this. Right now `setNoOp` `delete` the `op` key from the `op`. You want it to be updated to?

```javascript
function setNoOp(op) {
  op.op = null
  delete op.create;
  delete op.del;
}
```

alecgibson replied:
Yes, please. Just so that a no-op looks the same everywhere.

antoinelyset replied:
Oh okay. I try this change (and putting back `if (!server.op)`) and I've got some failing specs. I think it comes from the fact that I need to update this check in `_otApply` : 

https://github.com/antoinelyset/sharedb/blob/28f1faa7818d5af0ce9f610262a180f6651f6f61/lib/client/doc.js#L570

If I don't, `length` will fail here: 

https://github.com/sharedb/sharedb/blob/28f1faa7818d5af0ce9f610262a180f6651f6f61/lib/client/doc.js#L594

Right? But now we're back to the problem of `null` being skipped?

I'm a bit lost by this `setNoOp`, `null` is only a NoOp in JSON1 and ShareDB should not know this and continue as if it was a regular operation.

alecgibson replied:
That second link is broken but, I assume you meant to point [here](https://github.com/antoinelyset/sharedb/blob/28f1faa7818d5af0ce9f610262a180f6651f6f61/lib/client/doc.js#L594)? `length` shouldn't fail, because of the `defaultType` check (where the `defaultType` is always `json0`).

antoinelyset replied:
Yes sorry, it points here. the `op` can be a noOp and still be in `json0` no? This is what I have in the specs.

antoinelyset replied:
I've pushed my branch, if you want to take a look. I think I didn't quite understand.

alecgibson replied:
I think it's fine to update the check and just skip that block if `op.op === null`. All the block is doing is fragmenting the `json0` op and applying it one component at a time. Given that a no-op has no components, it's fine to just skip, and the no-op will be "applied" [below](https://github.com/antoinelyset/sharedb/blob/28f1faa7818d5af0ce9f610262a180f6651f6f61/lib/client/doc.js#L621).

antoinelyset replied:
Hmm I've update `_otApply` check for a `op.op === null` but now I've got a lot of other tests failing. I'm missing something again I think.

alecgibson commented on test/client/submit.js:
```diff
@@ -1196,5 +1198,34 @@ module.exports = function() {
       });
     });
+
+    describe('type.json1', function() {
```
Can we please move json1 tests into their own file for neatness?

alecgibson commented on test/client/submit.js:
```diff
@@ -1196,5 +1198,34 @@ module.exports = function() {
+          if (err) return done(err);
+          expect(doc.data).eql({age: 3});
+          // Version is 1 instead of 3, because the create and ops got composed
+          expect(doc.version).eql(1);
```
Can we break the `create` case into its own test?

alecgibson commented on test/client/submit.js:
```diff
@@ -1196,5 +1198,34 @@ module.exports = function() {
+              if (err) return done(err);
+              expect(doc.data).eql(undefined);
+              // del DOES NOT get composed
+              expect(doc.version).eql(4);
```
Again, separate test, please.

antoinelyset commented:
I've converted it to a draft. I will work on the other tasks (adding more tests related to the JSON1 format and to the server side)

antoinelyset commented:
Ok, so I've added a `submit-json1.js` test file, split the tests, and added one for server-side composition.

I've put back the change with `delete op.op` not because I'm stubborn (I can be sometimes hahaha), but because I don't quite know how to do it with `op.op = null`.

josephg commented:
We could add a isNoOp method in the json1, json0, etc and use that. They’re trivial methods to implement.

antoinelyset commented:
That could be interesting indeed. It could be done in another PR right?

curran commented:
Love it. `isNoOp` Sounds like a great thing.

alecgibson commented:
@antoinelyset what issues were you having using `op.op = null` instead of `delete op.op`? I think it's best to keep our "shape" of a no-op as consistent as possible.

antoinelyset commented:
I encountered a lot of errors linked to timeout and un-equalities. I think I didn't update enough places. (you can take a look here https://github.com/share/sharedb/pull/374/commits/ea40a735b30525e496bc23da4d027ad6f26b2cd7)

The way I think about it is that the no-op shape for ShareDB should be `delete op.op` . `op.op = null` is a no-op in JSON1 but it should be seen as a regular op by ShareDB. So we should keep the `delete op.op` and keep the updated checks `'op' in 'op'`.

If we want to avoid sending `op.op === null` in JSON1, we should implement a per type `isNoOp` check (`null` could be a valid and useful op in another type).

alecgibson commented:
@antoinelyset makes sense. We can always change it later if we need. The one other thing we mentioned in the meeting is that it's a bit weird that we check `'op' in op` but not `'create' in op` (I guess in theory there could be a no-op create...?), but that's probably just me nit-picking, so let's go ahead and merge.

Thanks for your work on this! 👏

curran commented:
This is great!

@alecgibson FWIW, the last commit in `master` seems to have broken CI.

![image](https://user-images.githubusercontent.com/68416/85720386-61daeb80-b6be-11ea-9aa2-974b80c6c67f.png)

Only Node 10.

![image](https://user-images.githubusercontent.com/68416/85720493-7b7c3300-b6be-11ea-8368-da0e65400f79.png)

Looks like it's a new test that's failing.

![image](https://user-images.githubusercontent.com/68416/85720638-9ea6e280-b6be-11ea-99f4-b09844991b83.png)

alecgibson commented:
Hopefully fixed in https://github.com/share/sharedb/pull/376
